### DISCUSSION TITLE
User Status Sync Between IDP and Keycloak

### DISCUSSION THREAD

**Discussion by AtulShu** [keycloak/keycloak#34076] [Q&A]
When an user account is disabled on a connected IPD, how do I ensure the user enabled flag is disabled automatically by syncing with IDP periodically. I looked for many approaches but not able to find a concrete solution.
<img width="138" alt="image" src="https://github.com/user-attachments/assets/5d1cf3c2-5b73-46d0-a6c6-45fba9238ec6">

It's ok for me if user stays in keycloak, I just want to disable it.
Can someone give some ideas on how to procede.

**Comment by danielFesenmeyer** [comment]
Afaik, Keycloak does not have a sync mechanism with IDPs, which runs periodically. Maybe you have a custom implementation for that, then this custom code can assure that this user property ("enabled") is preserved.

Keycloak provides a sync mechanism on login, which - at least per default - should not sync the "enabled" property. It it does, that's a bug from my POV.

**Reply by AtulShu** [reply]
Thanks for the reply, so just for my clarification: I need to make periodically API calls to connected IDP from keycloak and fetch user status and if it's disabled, I need to disable it in keycloak. Is there any other approach maybe through mappers or custom flow?

**Reply by danielFesenmeyer** [reply]
If it's ok when the Keycloak "enabled" status is changed on login (and login prevented when disabled), you could use a mapper. A precondition is that you can configure the IDP token to provide the "enabled" state as a claim.
But you probably need a custom mapper for that. The UserAttributeMapper almost does what you need, but it only supports "attributes" (like email, firstName, and user-defined attributes), not the "enabled" state.

**Comment by stianst** [comment]
Keycloak only sync changes from the IdP as part of the login, and not separately. We are at some point going to be looking into SCIM as a way to solve this use-case, where an IdP and Keycloak can be connected through SCIM.

There's another potential angle as well which is leveraging ID token and userinfo endpoints, to not just update the user as part of the login, but could be actively done through calling the userinfo endpoint when for example the ID token expires. Not something we support right now, but could perhaps be an interesting idea.
